### PR TITLE
Canonical potential, BPE and tokenization utils

### PR DESCRIPTION
**Pull request by yahya010**
- Canonicality potential that only allows canonical extensions of, uses `FastCanonicalityFilterBPE` for an efficient vectorized canonicality filter over next tokens
- Also includes option to add model-specific overrides (need to find them in each model), since there are very few edge cases for each model that the canonicality test misses
- Added `llist` dependency to pyproject.toml

### CODE REVIEW

File: genlm/control/potential/built_in/canonical.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        
+        Args:
+            tokenizer: The HuggingFace tokenizer to use
+            model_name: The name of the model (used for setting overrides)
+        """
+        from genlm.control.tokenization.bpe2 import FastCanonicalityFilterBPE
+        self.canonicality_filter = FastCanonicalityFilterBPE.from_huggingface(tokenizer)
+        self.canonicality_filter.set_overrides(model_name)
+        self.tokenizer = tokenizer
+        self.eos = EOS
```
Comment (benlebrun) [inline genlm/control/potential/built_in/canonical.py:26]: Not necessary since this attribute is added in the super class (also better not to hard code in case the super class logic changes).

**Comment by benlebrun** [comment]
This looks really awesome @yahya010! Thank you so much for this contribution. 

I added a few comments for some easy fixes. Note the failed linter check. 

As we've discussed, we probably want to reorganize the code; I think putting everything in the `canonical.py` file is the best solution for now. Eventually we can port a lot of this over to a dedicated tokenization library. 

One question: Do we know how the vocabulary decoding logic we have here interacts with the vocabulary decoding logic in `genlm-backend`? `PromptedLLM`s use the latter, so I am curious whether there are scenarios in which the vocabulary of a `PromptedLLM` and `Canonical` potential loaded using the same `model_name` will differ.

File: genlm/control/potential/built_in/canonical.py [hunk 2]
```diff
@@ -1,9 +1,345 @@
...
+    'ð', 'ñ', 'ò', 'ó', 'ô', 'õ', 'ö', '÷', 'ø', 'ù', 'ú', 'û', 'ü', 'ý', 'þ', 'ÿ',
+]
 
-class Canonical(Potential):
+# this is the inverse mapping of `_bytes_to_unicode`
+_decode_str_bytes = {s: i for i, s in enumerate(_encode_bytes_str)}
+_default_byte_decoder = _decode_str_bytes
+
+
+def decode_hf_tokenizer(tokenizer):
```
Comment (yahya010) [inline genlm/control/potential/built_in/canonical.py:34]: get merge rules function in potential, remove this (use the genlm.backend)

**Review by benlebrun — CHANGES_REQUESTED** [review]
Looks great @yahya010! I think there are still some things that we had discussed which need to be fixed though, see comments

File: genlm/control/potential/built_in/canonical.py [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+
+    def __call__(self, context):
+        if context == ():
+            mask = np.ones(self.V, dtype=bool)
+        else:
+            (_, last_token) = context
+            try:
+                left_id = self._encode[last_token]  # Get the ID of the last token
+            except KeyError:
+                # If last token is unknown, disallow everything except EOS, maybe raise error?
```
Comment (benlebrun) [inline genlm/control/potential/built_in/canonical.py:169]: I'd raise an error here

File: genlm/control/potential/built_in/canonical.py [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+            spine.reverse()
+            right_spine[i] = spine
+            max_width = max(max_width, len(spine))
+        return right_spine, max_width
+
+    def set_overrides(self, model_name):
+        if "gpt2" in model_name:
+            for left, right in [(198, 198), (2637, 82)]:
+                self.overrides[left].add(right)
+                warnings.warn(
```
Comment (benlebrun) [inline genlm/control/potential/built_in/canonical.py:251]: Might not actually be necessary to warn here. It probably won't be understandable what this means to the user

File: genlm/control/potential/built_in/canonical.py [hunk 5]
```diff
@@ -0,0 +1,453 @@
...
+        # Build _encode_byte (single byte -> token_id map)
+        _encode_byte = [None] * 256
+        for i in range(256):
+            byte_val = bytes([i])
+            if byte_val in _encode:
+                _encode_byte[i] = _encode[byte_val]
+
+        eos_token_ids = llm.token_maps.eos_idxs  # Assume this is already a list or set
+        if not eos_token_ids:
+            warnings.warn(
```
Comment (benlebrun) [inline genlm/control/potential/built_in/canonical.py:317]: I'd raise an error here if this actually has the effect of never allowing termination.
